### PR TITLE
Compute divergence

### PR DESCRIPTION
**Pull request by kidder**
## Proposed changes

Add function to compute divergence of a flux

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] Follows [code review guidelines](https://sxs-collaboration.github.io/spectre/code_review_guide.html)
- [ ] Code has documentation and unit tests
- [ ] Private member variables have a trailing underscore
- [ ] Do not use [Hungarian notation](https://en.wikipedia.org/wiki/Hungarian_notation), e.g. `double* pd_blah` is bad
- [ ] Header order:
  1. hpp corresponding to cpp (only in cpp files)
  2. Blank line (only in cpp files)
  3. STL and externals (in alphabetical order)
  4. Blank line
  5. SpECTRE includes (in alphabetical order)
- [ ] File lists in CMake are alphabetical
- [ ] Correct `noexcept` specification for functions (if unsure, mark `noexcept`)
- [ ] Mark objects `const` whenever possible
- [ ] Almost always `auto`, except with expression templates, i.e. `DataVector`
- [ ] All commits for performance changes provide quantitative evidence and the tests used to obtain said evidence.
- [ ] Make sure error messages are helpful, e.g. "The number of grid points in the matrix 'F' is not the same as the number of grid points in the determinant."
- [ ] Prefix commits addressing PR requests with `fixup`

### CODE REVIEW

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 1]
```diff
@@ -95,4 +95,13 @@ struct NormalDotNumericalFlux : db::DataBoxPrefix {
   using tag = Tag;
   static constexpr db::DataBoxString label = "NormalDotNumericalFlux";
 };
+
+/// \ingroup DataBoxTagsGroup
+/// \brief Prefix indicating the divergence
+template <typename Tag>
+struct div : db::DataBoxPrefix {
```
Comment (osheamonn) [inline src/DataStructures/DataBox/Prefixes.hpp]: Why is the casing different to the other prefix tags?

Comment (nilsdeppe) [reply src/DataStructures/DataBox/Prefixes.hpp]: It's the same as the `d` and `dt` prefix, however that doesn't mean we shouldn't change the capitalization of those

File: src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp [hunk 2]
```diff
@@ -24,6 +24,14 @@ struct LogicalImpl;
 /// \ingroup NumericalAlgorithmsGroup
 /// \brief Compute the partial derivatives of each variable with respect to
 /// the logical coordinate.
+///
+/// \requires `DerivativeTags` to be the head of `VariableTags`
+///
+/// \return a `Variables` with a spatial tensor index appended to the front
+/// of each tensor within `u` and each `Tag` wrapped with a `Tags::d`.
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp]: A note: I'm working on changing `Tags::d` to `Tags::deriv`, so maybe it's worth already making that change here so it doesn't get missed

File: src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp [hunk 3]
```diff
@@ -34,6 +42,14 @@ auto logical_partial_derivatives(const Variables<VariableTags>& u,
 /// \ingroup NumericalAlgorithmsGroup
 /// \brief Compute the partial derivatives of each variable with respect to
 /// the coordinates of `DerivativeFrame`.
+///
+/// \requires `DerivativeTags` to be the head of `VariableTags`
+///
+/// \return a `Variables` with a spatial tensor index appended to the front
+/// of each tensor within `u` and each `Tag` wrapped with a `Tags::d`.
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp]: Same note

File: src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp [hunk 4]
```diff
@@ -24,6 +24,14 @@ struct LogicalImpl;
 /// \ingroup NumericalAlgorithmsGroup
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp:57]: Commit message: "Document template parameter of logical and partial_derivatives"

File: src/DataStructures/DataBox/Prefixes.hpp [hunk 5]
```diff
@@ -95,4 +95,13 @@ struct NormalDotNumericalFlux : db::DataBoxPrefix {
   using tag = Tag;
   static constexpr db::DataBoxString label = "NormalDotNumericalFlux";
 };
+
+/// \ingroup DataBoxTagsGroup
+/// \brief Prefix indicating the divergence
+template <typename Tag>
```
Comment (nilsdeppe) [inline src/DataStructures/DataBox/Prefixes.hpp]: This tag won't work as a compute item, I'm working on a PR for `Tags::deriv` that does this so we will want to update `div` to be the same

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+///
+/// \return a `Variables` with the same structure as `F` and each `Tag` wrapped
+/// with a `Tags::div`.
+template <typename FluxTags, size_t Dim, typename DerivativeFrame>
+auto divergence(
+    const Variables<FluxTags>& F, const Index<Dim>& extents,
+    const Tensor<DataVector, tmpl::integral_list<std::int32_t, 2, 1>,
+                 typelist<SpatialIndex<Dim, UpLo::Up, Frame::Logical>,
+                          SpatialIndex<Dim, UpLo::Lo, DerivativeFrame>>>&
+        inverse_jacobian) noexcept {
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: Return type can't be auto for compute item. I used trailing return type in partial_derivatives

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp [hunk 7]
```diff
@@ -0,0 +1,238 @@
...
+#include "NumericalAlgorithms/Spectral/LegendreGaussLobatto.hpp"
+#include "Utilities/MakeArray.hpp"
+#include "tests/Unit/TestHelpers.hpp"
+
+namespace {
+
+template <size_t Dim>
+struct Flux1 : db::DataBoxTag {
+  using type = tnsr::I<DataVector, Dim, Frame::Grid>;
+  static constexpr db::DataBoxString label = "Vector_t";
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp]: I'm confused by this label

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp [hunk 8]
```diff
@@ -0,0 +1,238 @@
...
+      result.get(i) += term;
+    }
+    return result;
+  }
+};
+
+template <size_t Dim>
+struct Flux2 : db::DataBoxTag {
+  using type = tnsr::Ij<DataVector, Dim, Frame::Grid>;
+  static constexpr db::DataBoxString label = "Vector_t";
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp]: And this label

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp (hunk 6 above)
Comment (osheamonn) [reply src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: Does it need to be trailing? The return type is `Variables<db::wrap_tags_in<Tags::div, FluxTags>>` which I thought would be deducible.

Comment (nilsdeppe) [reply src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: I don't think it has to be, no, either way is fine to me

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 9]
```diff
@@ -0,0 +1,51 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines functions computing partial derivatives.
```
Comment (wthrowe) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: Nope.

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp [hunk 10]
```diff
@@ -0,0 +1,238 @@
...
+    tmpl::for_each<FluxTags>([&a, &x, &expected_div_F ](auto tag) noexcept {
+      using FluxTag = typename decltype(tag)::type;
+      using DivFluxTag = Tags::div<FluxTag>;
+      get<DivFluxTag>(expected_div_F) = FluxTag::div({{a}}, x);
+    });
+
+    const auto div_F = divergence<FluxTags>(F, extents, inverse_jacobian);
+
+    for (size_t n = 0; n < div_F.size(); ++n) {
+      CAPTURE_PRECISE(div_F.data()[n] - expected_div_F.data()[n]);  // NOLINT
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp]: Comment what is being NOLINTed.

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp (hunk 6 above)
Comment (kidder) [reply src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: please clarify: auto is okay, or I need to explicitly give the return type?

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp (hunk 7 above)
Comment (kidder) [reply tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp]: I am too

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp (hunk 9 above)
Comment (kidder) [reply src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: Doh!

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp (hunk 6 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: You need to specify the return type

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp (hunk 7 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp]: I found out while writing tests for the partial derivative compute item that these were just copied from the partial derivative test. I've changed that in #415 for the partial derivative test, not sure why that was there originally, but whatever :)

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp (hunk 6 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: This is now visible in #415 but there is a discussion about where to put the tags and whether or not to generate explicit instantiations functions that take Variables

**Comment by nilsdeppe** [comment]
This should now be able to be updated to follow the style of partial derivatives.

File: src/Utilities/StdHelpers.hpp [hunk 11]
```diff
@@ -401,6 +401,19 @@ inline std::array<T, Dim - 1> all_but_specified_element_of(
   return result;
 }
 
+/// \ingroup UtilitiesGroup
+/// \brief Construct an array from an existing array prepending a value
+template <typename T, size_t Dim>
+inline std::array<T, Dim + 1> prepend(const std::array<T, Dim>& a,
```
Comment (kidder) [inline src/Utilities/StdHelpers.hpp]: constexpr

File: src/Utilities/StdHelpers.hpp [hunk 12]
```diff
@@ -401,6 +401,19 @@ inline std::array<T, Dim - 1> all_but_specified_element_of(
...
+                                      const T& value) noexcept {
```
Comment (kidder) [inline src/Utilities/StdHelpers.hpp]: take value by value and move it in

File: src/DataStructures/Tensor/Metafunctions.hpp [hunk 13]
```diff
@@ -605,4 +605,13 @@ using prepend_spacetime_index = Tensor<
                            tmpl::max<tmpl::_state, tmpl::_element>>::value>>,
     tmpl::push_front<typename Tensor_t::index_list,
                      SpacetimeIndex<VolumeDim, Ul, Fr>>>;
+
+/// \ingroup TensorGroup
+/// \brief remove the first index of a tensor
+/// \tparam Tensor_t the tensor type whose first index is removed
+template <typename Tensor_t>
```
Comment (kidder) [inline src/DataStructures/Tensor/Metafunctions.hpp]: remove _t, promote Tensor to ::Tensor

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 14]
```diff
@@ -0,0 +1,105 @@
...
+/// \brief Prefix indicating the divergence
+///
+/// There are three variants of this tag that change how the derivatives are
+/// computed depending on which is chosen. The two simplest are the non-compute
+/// item versions for Tensor and Variables tags. These take two template
+/// parameters:
+/// 1. The tag to wrap
+/// 2. The frame of the derivative index in `Tensor`s
+///
+/// The third variant of the derivative tag is a compute item that computes
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: derivative -> divergence

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 15]
```diff
@@ -0,0 +1,105 @@
...
+/// the Tag for inverse Jacobian between the logical frame and the frame in
+/// which the derivatives are taken.
+template <typename T, typename S>
+struct div : Tags_detail::div_impl<T, S> {};
+}  // namespace Tags
+
+/// \ingroup NumericalAlgorithmsGroup
+/// \brief Compute the (Euclidean) divergence of fluxes
+///
+/// \return a `Variables` with the same structure as `F` and each `Tag` wrapped
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: `Tag` -> tag in `FluxTags`

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+      tmpl::back<typename db::item_type<InverseJacobianTag>::index_list>;
+  using flux_tags = tmpl::list<FluxTags...>;
+
+ public:
+  static constexpr db::DataBoxString label = "div";
+  static constexpr auto function =
+      divergence<flux_tags, derivative_frame_index::dim,
+                 typename derivative_frame_index::Frame>;
+  using argument_tags =
+      typelist<Tags::Variables<flux_tags>,
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: typelist -> tmpl::list

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 17]
```diff
@@ -0,0 +1,105 @@
...
+                 typelist<SpatialIndex<Dim, UpLo::Up, Frame::Logical>,
+                          SpatialIndex<Dim, UpLo::Lo, DerivativeFrame>>>&
+        inverse_jacobian) noexcept;
+
+namespace Tags {
+namespace Tags_detail {
+template <typename Tag, typename Frame>
+struct div_impl<Tag, Frame, Requires<tt::is_a_v<Tensor, db::item_type<Tag>>>>
+    : db::DataBoxPrefix {
+  using type = TensorMetafunctions::remove_first_index<db::item_type<Tag>>;
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: static_assert that the first index is in Frame

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 18]
```diff
@@ -0,0 +1,105 @@
...
+  using tag = Tag;
+  static constexpr db::DataBoxString label = "div";
+};
+
+template <typename Tag, typename Frame>
+struct div_impl<Tag, Frame,
+                Requires<tt::is_a_v<::Variables, db::item_type<Tag>>>>
+    : db::DataBoxPrefix {
+  using type = db::item_type<Tag>;
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: fix this (wrap_tags)

File: src/NumericalAlgorithms/LinearOperators/Divergence.hpp [hunk 19]
```diff
@@ -0,0 +1,105 @@
...
+  using tag = Tag;
+  using frame = Frame;
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.hpp]: remove

File: src/NumericalAlgorithms/LinearOperators/Divergence.cpp [hunk 20]
```diff
@@ -0,0 +1,51 @@
...
+
+#include "NumericalAlgorithms/LinearOperators/Divergence.hpp"
+
+#include "DataStructures/Index.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "NumericalAlgorithms/LinearOperators/PartialDerivatives.hpp"
+#include "Utilities/StdHelpers.hpp"
+
+/// \ingroup NumericalAlgorithmsGroup
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.cpp]: remove

File: src/NumericalAlgorithms/LinearOperators/Divergence.cpp [hunk 21]
```diff
@@ -0,0 +1,51 @@
+// Distributed under the MIT License.
```
Comment (kidder) [inline src/NumericalAlgorithms/LinearOperators/Divergence.cpp]: rename file with tpp

File: tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp [hunk 22]
```diff
@@ -0,0 +1,257 @@
...
+  }
+};
+
+template <size_t Dim, typename Frame>
+using two_fluxes = typelist<Flux1<Dim, Frame>, Flux2<Dim, Frame>>;
+
+template <size_t Dim, typename Frame = Frame::Inertial>
+void test_divergence(
+    const Index<Dim>& extents,
+    std::array<std::unique_ptr<MathFunction<1>>, Dim>&& functions) noexcept {
```
Comment (kidder) [inline tests/Unit/NumericalAlgorithms/LinearOperators/Test_Divergence.cpp]: take by value

File: src/Utilities/StdHelpers.hpp [hunk 23]
```diff
@@ -404,10 +404,10 @@ inline std::array<T, Dim - 1> all_but_specified_element_of(
 /// \ingroup UtilitiesGroup
 /// \brief Construct an array from an existing array prepending a value
 template <typename T, size_t Dim>
-inline std::array<T, Dim + 1> prepend(const std::array<T, Dim>& a,
-                                      const T& value) noexcept {
+inline constexpr std::array<T, Dim + 1> prepend(const std::array<T, Dim>& a,
+                                                const T value) noexcept {
```
Comment (nilsdeppe) [inline src/Utilities/StdHelpers.hpp]: remove const

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
squash

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Pull request merged**